### ISSUE TITLE
TC search treats `match` (`if`) as TC-opaque

### ISSUE THREAD

**Issue by Janno** [coq/coq#13209]
#### Description of the problem

In the following example, only the first and third calls to `apply _` succeed. It is hard to say which calls should succeed but I think it is rather reasonable to say that the current behaviour is surprising.

```coq
Class A (n : nat) := {}.
Definition three := 3.
Instance A_three : A three := {}.
Definition three' := if true then three else 1.
(* 1 *) Goal A (if true then three else 1). apply _. Qed.
Hint Opaque three : typeclass_instances.
(* 2 *) Goal A (if true then three else 1). Fail apply _. Abort.
(* 3 *) Goal A (three'). apply _. Qed.
```

Here is my interpretation of what Coq is doing:
1) Coq sees a TC-opaque `if` term and a TC-transparent constant `three` and it given the usual rule of TC search Coq optimistically tries to unify them since at least one of them is TC-transparent.
2) Coq sees a TC-opaque `if` term in the goal and an instance on the TC-opaque constant `three` and concludes that it should not attempt to unify these two because both are TC-opaque.
3) Coq sees a TC-transparent constant `three'` and a TC-opaque constant `three` and attempts to unify them since one of them is TC-transparent.

So basically this makes total sense if you buy the assumption that `if` (i.e. `match`) terms should be TC-opaque. However, given the counter-intuitive consequences of this choice and the fact that basically anything else is considered transparent by default I wonder if it wouldn't make more sense to make `match` terms TC-transparent by default at least for now? Alternatively (but also independently) maybe Coq should have a `Hint Match Transparent/Opaque` command?

There is some discussion about this in zulip: https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/Opaque.20does.20not.20imply.20Typeclasses.20Opaque

#### Coq Version
8.11.2

**Comment by RalfJung** [comment]
> So basically this makes total sense if you buy the assumption that if (i.e. match) terms should be TC-opaque. 

I do not understand how that is the conclusion you reached here? The first goal (1) does unify an `if` with `three`, so clearly they are *not* treated as TC-opaque?

What I do not understand is how making `three` opaque (the difference between (1) and (2)) affects unifying `if true then three else 1` with `three` -- whether these two unify should have nothing to do with opaqueness of `three` as `three` does not need to be unfolded to perform this unification.

**Comment by Janno** [comment]
> I do not understand how that is the conclusion you reached here? The first goal (1) does unify an if with three, so clearly they are not treated as TC-opaque?

There are two stages here: one is the discrimination net that decides what could possibly work and one is the unification that happens after that. Here's the TC debug output for all three examples:

```
Debug: 1: looking for (A (if true then three else 1)) without backtracking
Debug: 1.1: exact A_three on (A (if true then three else 1)), 0 subgoal(s)

Debug: 1: looking for (A (if true then three else 1)) without backtracking
Debug: 1: no match for (A (if true then three else 1)), 0 possibilities

Debug: 1: looking for (A three') without backtracking
Debug: 1.1: exact A_three on (A three'), 0 subgoal(s)
```

Note that in the second example there are *zero* possibly applicable instances. My conclusion from that is that the discrimination net considers the `if` term opaque and thus rules out any possibility of it unifying with `three` (since that is marked TC opaque in this goal).

> What I do not understand is how making three opaque (the difference between (1) and (2)) affects unifying if true then three else 1 with three -- whether these two unify should have nothing to do with opaqueness of three as three does not need to be unfolded to perform this unification.

I really do think that the reason example (1) succeeds is because `three` is TC transparent and that means Coq cannot rule out the possibility that it will reduce to/unify with the `if` term modulo TC opacity. Of course we know that cannot happen. But Coq is optimistic and because the `if` terms are not actually opaque for TC *unification* the whole thing still works.. for all the wrong reasons.

**Comment by RalfJung** [comment]
> Here's the TC debug output for all three examples:

Are you sure that this will show attempts that fail due to unification? What does it say with higher debug level?

> My conclusion from that is that the discrimination net considers the if term opaque 

Oh I see. Somehow my internal model of this says that `Typeclasses Opaque` is only about unification and that the "discrimination net" always works by direct syntactic matching. I am not sure if that is actually true.

But anyway isn't the discrimination net only about the top-level part, i.e., `A` here? What does your mental model say for how much information the discrimination net is taking into account?

**Comment by Janno** [comment]
> Are you sure that this will show attempts that fail due to unification? What does it say with higher debug level?

Those attempts wouldn't be *shown* but they would be counted in `x possibilities`. Here's the output at the highest debug level:

```
Debug: Starting resolution with 1 goal(s) under focus and 0 shelved goal(s) in only_classes mode, unbounded
Debug: 1: looking for (A (if true then three else 1)) without backtracking
Debug: 1.1: exact A_three on (A (if true then three else 1)), 0 subgoal(s)
Debug: 1.1: after exact A_three finished, 0 goals are shelved and unsolved ( )
Debug: Starting resolution with 1 goal(s) under focus and 0 shelved goal(s) in only_classes mode, unbounded
Debug: 1: looking for (A (if true then three else 1)) without backtracking
Debug: 1: no match for (A (if true then three else 1)), 0 possibilities
Debug: Starting resolution with 1 goal(s) under focus and 0 shelved goal(s) in only_classes mode, unbounded
Debug: 1: looking for (A three') without backtracking
Debug: 1.1: exact A_three on (A three'), 0 subgoal(s)
Debug: 1.1: after exact A_three finished, 0 goals are shelved and unsolved ( )
```

> But anyway isn't the discrimination net only about the top-level part, i.e., A here? What does your mental model say for how much information the discrimination net is taking into account?

I think the discrimination net considers terms up to a certain depth, IIRC. It's not too hard to test by throwing more complex terms at TC search.. but I don't feel inspired right now. @ppedrot can give you an authoritative answer, I think.

**Comment by RalfJung** [comment]
Okay, I once again realize that my mental model for the TC search algorithm is wrong then.^^ AFAIK that algorithm is also not precisely described anywhere, is it?

**Comment by Janno** [comment]
It seems this issue is a tiny bit more complex than I thought initially. There is an asymmetry: goals with `ifs` only match `_` hint patterns, but hints keyed on `ifs` match any goal (because the `if` term is translated to `_`). I hope the following example (dual to the original one) illustrates this. (I removed `three'` because it is not relevant here, I think.)
```coq
Class A (n : nat) := {}.
Definition three := 3.
Instance A_three_if : A (if true then three else 1) := {}.
(* 1 *) Goal A three. apply _. Qed.
Hint Opaque three : typeclass_instances.
(* 2 *) Goal A three. apply _. Qed.  (* This combination fails in the original example *)
```

While I cannot claim to understand the code, I think there is a good chance that this behavior can be read off of btermdn.ml: in patterns, the catch-all case (that encompasses case nodes) maps the pattern to `None`, i.e. `_`. However, during lookups, the catch-all case maps `Case _` to `Nothing`. I imagine (but cannot find any proof) that `Nothing` matches nothing except `_`, which would explain why only hints with `_` patterns match `if` queries. Conversely, instances keyed on `if` will of course match all queries because the corresponding patterns are always translated to `_`.

Because of this I no longer think that adding a flag is enough to make the whole story around `ifs` make sense. 


> Okay, I once again realize that my mental model for the TC search algorithm is wrong then.^^ AFAIK that algorithm is also not precisely described anywhere, is it?

I don't think so. I am slowly making progress by reading the code it in lieu of documentation. I usually use it as starting point and then bug the coq developers on zulip.
